lib/graphql/analysis/ast/visitor.rb

```diff
@@ -0,0 +1,228 @@
+# frozen_string_literal: true
+module GraphQL
+  module Analysis
+    module AST
+      class Visitor < GraphQL::Language::Visitor
+        def initialize(query:, analyzers:)
+          @analyzers = analyzers
+          @path = []
+          @object_types = []
+          @directives = []
+          @field_definitions = []
+          @argument_definitions = []
+          @directive_definitions = []
+          @query = query
+          @schema = query.schema
+          @response_path = []
+          @skip_stack = [false]
+          super(query.document)
+        end
+
+        attr_reader :context, :query
+
+        # @return [Array<GraphQL::ObjectType>] Types whose scope we've entered
+        attr_reader :object_types
+
+        # @return [Array<String>] The nesting of the current position in the AST
+        def path
+          @path.dup
+        end
+
+        def arguments_for(ast_node, field_definition)
+          @query.arguments_for(ast_node, field_definition)
+        end
+
+        def on_operation_definition(node, parent)
+          object_type = @schema.root_type_for_operation(node.operation_type)
+          @object_types.push(object_type)
+          @path.push("#{node.operation_type}#{node.name ? " #{node.name}" : ""}")
+          call_analyzers(:on_enter_operation_definition, node, parent)
+          super
+          call_analyzers(:on_leave_operation_definition, node, parent)
+          @object_types.pop
+          @path.pop
+        end
```

Comment:
This reminds me of a bug in the current analyzer code which we can fix on this one, in a later PR maybe. 

Currently we visit _all_ operation definitions. But we should probably only visit `query.selected_operation`.

Reply:
👍 fixed in this PR.